### ISSUE TITLE
Optimize OverlappingFieldsCanBeMerged

### ISSUE THREAD

**Issue by objmagic** [sangria-graphql/sangria#296]
`QueryReducer.rejectComplexQueries` become stuck with a 412KBytes GraphQL query selecting ~150000 fields. Did a profiling and stacktrace shows we are doing tremendous amount of work in `OverlappingFieldsCanBeMerged`

An example of the query is at [here](https://gist.github.com/objmagic/3c881449fcdb3a812a371b86bfa5a3c9)

We are using:

* **Sangria version 1.3.2**
* **JVM 8 (with some Twitter internal customization)**

A screenshot of part of the stacktrace (sorry I cannot provide complete stacktrace).

![image](https://user-images.githubusercontent.com/3656079/33035178-14f4d5e4-cddf-11e7-8077-a087137da451.png)

**Comment by objmagic** [comment]
@OlegIlyenko 

Thanks a lot for your prompt reply. I would love to work through this issue with you.

Let me see if I can get a better stacktrace.

**Comment by OlegIlyenko** [comment]
This would be awesome! 👍 As a first step, I think it would be really helpful to verify that mentioned validation is indeed the bottleneck. Can you please exclude it in your tests:

```scala
QueryReducerExecutor.reduceQueryWithoutVariables(
  ...
  queryReducers = ...,
  queryValidator = 
    new RuleBasedQueryValidator(
      QueryValidator.allRules
        .filterNot(_.isInstanceOf[OverlappingFieldsCanBeMerged])))

// or

Executor.execute(
  ...
  queryReducers = ...,
  queryValidator = 
    new RuleBasedQueryValidator(
      QueryValidator.allRules
        .filterNot(_.isInstanceOf[OverlappingFieldsCanBeMerged])))
```

If we can verify that it cases the slowdown, then we can focus on `OverlappingFieldsCanBeMerged`.

**Comment by objmagic** [comment]
remove the `OverlappingFieldsCanBeMerged` validator solves the issue.

**Comment by OlegIlyenko** [comment]
Great, this means that we can isolate the issue and focus the effort on optimising the `OverlappingFieldsCanBeMerged` validation. Maybe you could also change the issue title to reflect this fact?

If you would like to investigate the issue as well, this here are some usefule resources:

* The validation implementation: [OverlappingFieldsCanBeMerged](https://github.com/sangria-graphql/sangria/blob/master/src/main/scala/sangria/validation/rules/OverlappingFieldsCanBeMerged.scala)
* The tests: [OverlappingFieldsCanBeMergedSpec](https://github.com/sangria-graphql/sangria/blob/master/src/test/scala/sangria/validation/rules/OverlappingFieldsCanBeMergedSpec.scala)
* Reference impl (sangria implementation is very close to it, but has a recent bugfix which is not yet in the ref. impl. - this might case some perf issues as well): [OverlappingFieldsCanBeMerged.js](https://github.com/graphql/graphql-js/blob/master/src/validation/rules/OverlappingFieldsCanBeMerged.js) (be careful, reference impl. still contains this bug: https://github.com/graphql/graphql-js/pull/780)

**Comment by OlegIlyenko** [comment]
Now that I think about it, recent bugfix introduced a change that reduces usage of the cache. i described this issue here: https://github.com/graphql/graphql-js/pull/780#pullrequestreview-49855652

This might be a promising path to investigate. For example, we can revert the bugfix and see how it perform without it. This commit introduced the bugfix: https://github.com/sangria-graphql/sangria/commit/9866a7bfab8d840185b4d646ee8909e9b9882749#diff-9ce8420e3a37d01d401ac1bc89e17e6e

**Comment by OlegIlyenko** [comment]
The validation rules are normally completely self-contained. So you can just copy/paste [this old version of the validation](https://github.com/sangria-graphql/sangria/blob/5f6279b4a810b3674f53d579e43897c48de590bc/src/main/scala/sangria/validation/rules/OverlappingFieldsCanBeMerged.scala) under a different name in your test and add it in the `RuleBasedQueryValidator` instead of the new version of `OverlappingFieldsCanBeMerged`.

**Comment by objmagic** [comment]
Thanks. I'll try to find some time to debug this.

**Comment by objmagic** [comment]
Notice this problem can introduce serious problem. When doing `curl` on our API server, it takes 8~10 mins to get back a bad result. That said, attacker construct maliciously large query to DDoS your production service.

**Comment by OlegIlyenko** [comment]
Indeed, this is a concern. 

Assuming that parsing and `OverlappingFieldsCanBeMerged` as well as other validations are well optimized. If incoming request is allowed to have unlimited payload size, then eventually parsing and validation become too slow, it is just a matter of payload size. 

So far I thought about following possible solutions that generally implemented outside of GraphQL execution:

* Always limit the request payload size
* Use persisted queries, if possible
* Track execution time and cancel the execution if it takes more time than a threshold (this can be also limited in scope, e.g. only parsing + validation). This is actually something that library can make easier to do. Since parsing and validation are CPU-bound tasks, it should be relatively easy to add simple cancellation mechanism (it becomes harder when IO with external systems is involved).

I would be very interested to hear you opinion and ideas on possible solutions and countermeasures for this issue.

**Comment by objmagic** [comment]
For now, we added a simple HTTP filter to filter out request that is above certain threshold.

**Comment by objmagic** [comment]
![image](https://user-images.githubusercontent.com/3656079/33462168-f61c3d56-d5ea-11e7-96bd-8e146474df05.png)

@OlegIlyenko using old version of the validation file and got the result above. The situation is still bad but I got part of the result back and then got bunch of HTTP timeout failures. That said, it seems the bug fix does not matter much here.

**Comment by OlegIlyenko** [comment]
I see, thanks for investigating this possibility! I guess we need to look at the original algorithm and try to optimise it.

**Comment by objmagic** [comment]
Thanks. I wouldn't worry much now since we have put an HTTP filter at the very front, and all incoming traffic will be validated before reaching DCs. 

Let's keep this issue open, just for other people who might have the same concern.

**Comment by SimonAdameit** [comment]
@objmagic Does that HTTP filter abort the query validation or does it still hog resources on the server and just timeout the client?

We've had a go at optimising this validation, but did not come up with a full solution. As others may come upon that also, I'll document some of the ideas here.

The current algorithms does O(n^2) pairwise comparisons for fields that can potentially conflict due to the same output name. This means already pretty simple queries can lead to a complexity explosion. E.g.:
```
query {
  name
  name 
  name
  name
  ....
}
```
The number of comparisons done here will be `n choose 2`. To simplify that we thought we could have a more granular bucketing of fields than just the output name. I.e.:
```
output_name ->          // potential conflicts
  (output_type ->       //  has to be unique and is also accounted for when parent types are exclusive
    (exclusive_parent_types -> // unfortunately open vs closed types require double entries
      (field_name_and_args ->  // has to be unique on this level
        field_bucket)))
```
Using such a scheme we could already significantly improve the above query (linear vs. quadratic growth). Unfortunately that required inlining fragments into these buckets, and thus introduced another source of complexity.

After sketching some further ideas to reduce the complexity we currently stopped that effort.

An alternative for us would be to introduce a budget for the execution in the current algorithm.

**Comment by SimonAdameit** [comment]
We also experienced this issue at [XING](https://www.xing.com) and invested some time to fix it. Here is the PR: https://github.com/sangria-graphql/sangria/pull/458 and here is the [description of the algorithm](https://tech.xing.com/graphql-overlapping-fields-can-be-merged-fast-ea6e92e0a01).

**Comment by yanns** [comment]
The faster version is now the default one starting from https://github.com/sangria-graphql/sangria/releases/tag/v3.0.0
